### ISSUE TITLE
The header must be fetched from jakarta.ee

### ISSUE THREAD

**Issue by ivargrimstad** [eclipse-ee4j/starter#246]
The header (with menu), as well as the footer, must be synced with the one at https://jakarta.ee

This is the Starter header today:
<img width="1181" alt="starter-INCORRECT" src="https://user-images.githubusercontent.com/149188/229049452-a5c2515e-8fb4-4c76-a5fc-8063c6b95c5b.png">

This is how it should be:
<img width="1235" alt="jakarta ee-CORRECT" src="https://user-images.githubusercontent.com/149188/229049459-c5f12ea1-bd1d-441f-a741-4639101a14ec.png">

The sources for the https://jakarta.ee website can be found here: https://github.com/jakartaee/jakarta.ee
It is built with Hugo, but should it should be possible to include the header and footer in Faces.

**Comment by kito99** [comment]
@ivargrimstad is it okay to copy it or are you expecting it to be pulled dynamically during the build?

**Comment by m-reza-rahman** [comment]
There should be a job setup shortly, so if we want to do something at build time, it should be doable. @jeyvison can help you with the setup,

**Comment by kito99** [comment]
@ivargrimstad correct me if I'm wrong, but wouldn't we need to create a separate yarn run script that builds the header by itself? https://github.com/jakartaee/jakarta.ee/blob/src/layouts/partials/header.html. 

Since this is using Hugo, there is no static version of the header we can just include.

**Comment by BalusC** [comment]
Public instructions how to reuse the jakarta.ee header from its original source in any random website would be helpful. A dedicated API service which returns the desired raw HTML would be absolutely fabuluous.

**Comment by kito99** [comment]
@ivargrimstad is there someone who's familiar with the web site build (and hugo) that can help us out with this? As @BalusC  mentioned, it's really not clear how to import it dynamically. We can always cut and paste from the generated site, but that is of course prone to long term issues.

**Comment by ivargrimstad** [comment]
> @ivargrimstad is there someone who's familiar with the web site build (and hugo) that can help us out with this? As @BalusC mentioned, it's really not clear how to import it dynamically. We can always cut and paste from the generated site, but that is of course prone to long term issues.

@chrisguindon Do you think someone in webdev could help with this?

**Comment by chrisguindon** [comment]
@ivargrimstad We might be able to help. Is it correct for me to assume that need a url/page from jakarta.ee website that only includes the HTML for the footer and header?


@autumnfound can you help here?

**Comment by autumnfound** [comment]
> @ivargrimstad We might be able to help. Is it correct for me to assume that need a url/page from jakarta.ee website that only includes the HTML for the footer and header?
> 
> @autumnfound can you help here?

I can poke around and see about making some raw HTML fragments that can be generated per site, but that comes with a caveat. There would be no real way to customize the metadata of the page easily, as it would go through the static generator. I could add placeholders that are easy to identify at least, but you'd need to use string substitution to replace those. This is a limitation of the unique combination of systems of automated export of this, and the fact that our sites are statically rendered.

**Comment by autumnfound** [comment]
> If it were exposed as an API service, these could be supplied as parameters.

There are no current plans to implement this as an API, but there is a different solution we can offer that help resolve this need. I've created a [patch to our theme](https://gitlab.eclipse.org/eclipsefdn/it/webdev/hugo-solstice-theme/-/merge_requests/306) that will allow for sites that add the required content to supply HTML fragments for the header and footer. These are pretty simple fragments that include the header w/o breadcrumbs, and the full standard footer.

For the header fragment ([preview from the PR](https://preview-306--webdev-eclipse-org-docs-hugo.eclipsecontent.org/docs/hugo/templates/header/)), I've included some easily identified placeholders to make basic customization possible:

- `{{TITLE}}` for page title
- `{{KEYWORDS}}` for SEO keywords
- `!!DESCRIPTION!!` for page description. Due to a limitation in how we print this content, I needed to use a less unique pattern than above.

For the footer fragment ([preview from the PR](https://preview-306--webdev-eclipse-org-docs-hugo.eclipsecontent.org/docs/hugo/templates/footer/)), there are no default placeholders at the moment, though I can include a placeholder for importing custom JS or other assets at the bottom of the document.

Does this cover what you would need to get started?

**Comment by autumnfound** [comment]
> @autumnfound ah, okay. That sounds prefect. How do we parameterize the template? URL parameters?

You'd have to use string replace for the placeholders I mentioned above. As this is a static site, there is no way to parameterize this, unfortunately.

**Comment by kito99** [comment]
@autumnfound ok, now I get it. this would work perfectly! however, I think we'd need this for the jakarta.ee site rather than the eclipse.org site.

**Comment by kito99** [comment]
@autumnfound bump 🙂

**Comment by autumnfound** [comment]
Apologies, we had our annual all hands last week and this slipped attention. We need to update the theme in the main site, and then I can get the template up for Jakarta.ee. I can look into doing it next week as there are some holidays coming up.

**Comment by autumnfound** [comment]
Just to let you guys know, the patch work has been started, but we uncovered some stuff we want to move to our shared front end code, so it's paused while we get that done and published. We're aiming to get this out this week if possible still though

https://github.com/jakartaee/jakarta.ee/pull/1686

**Comment by autumnfound** [comment]
Alright, the initial version is now available! See https://jakarta.ee/templates/header/ and https://jakarta.ee/templates/footer/ for the fragments

**Comment by kito99** [comment]
Thanks @autumnfound! This is just what we need. We can just grab this during the build process. I noticed the menu links don't work; is this something you're planning to work on?

**Comment by chrisguindon** [comment]
>  I noticed the menu links don't work; is this something you're planning to work on?

The header links should work if you have both the header and footer on the same page. The required .js file should be in right before the closing `</body>` tag.

**Comment by chrisguindon** [comment]
However, I am noticing that the CSS and js files paths don't include the full domain. 

It will need to point to jakarta.ee/ for them to work outside of the jakarta.ee domain:

```
<script src="/js/solstice.js?v2.3?v=1688650700)"></script>
<link rel="stylesheet" href="/css/styles.css?v2.3?v=1688650700">
```

**Comment by kito99** [comment]
@chrisguindon makes sense. thanks so much! I'll let you know when I'm able to test it out.

**Comment by m-reza-rahman** [comment]
Hi @kito99, any progress on this?

**Comment by kito99** [comment]
@m-reza-rahman not yet, but I can make it a priority and get to it in late Oct / early Nov.

**Comment by m-reza-rahman** [comment]
Thanks as always Kito! A quick question if you don’t mind. I have a Google Summer of Code student that says he is keen on working on the look and feel. He is enthusiastic but inexperienced. Do you think you have the bandwidth to mentor him and get him to do the work? Or is it best that you handle this one on your own?

**Comment by kito99** [comment]
Hey Reza,

I don’t have a lot of bandwidth but I’d love to do it. I’ve been eager to mentor younger folks more.

___

Kito D. Mann | @kito99 | Java Champion | Google Developer Expert Alumni | LinkedIn
Expert consulting and training: Cloud architecture and modernization, Java/Jakarta EE, Web Components, Angular, Mobile Web
Virtua, Inc. | virtua.tech
+1 203-998-0403

* Enterprise development, front and back. Listen to Stackd Podcast.
* Speak at conferences? Check out SpeakerTrax.

**Comment by m-reza-rahman** [comment]
Great! Let me make introductions.

**Comment by m-reza-rahman** [comment]
Hi @kito99, I tried to connect you to Yash via email, but the email I have for you bounced. Can you please share your updated email with me privately?

**Comment by kito99** [comment]
@yashTEF an I discussed this today, and this is our current plan:

1. Replace the existing  `topbar.xhtml` and`footer.xhtml` and  files with the new header (https://jakarta.ee/templates/header/) and footer (https://jakarta.ee/templates/footer/) content. This will make it easy to ensure things are working correctly. 

@yashTEF, you should only use the contents of the `<body>` of the documents, and add the stylesheets to the `WEB-INF\tempate.xhtml` file.

2. During the build process, grab the most recent versions of the header and footer from the URLs above (extracting only the `<body>`) and copy them into the existing `footer.xhtml` and `topbar.xhtml` files (presumably using the Maven resources plugin (https://maven.apache.org/plugins/maven-resources-plugin/). We haven't decided how exactly to grab the files and perform that manipulation (`curl` and unix commands in general come to mind, but there may be maven plugins that do this.)

This allows us to have a basic working version of the header/footer in the repo, but updated it dynamically during the build.

@ivargrimstad  or @m-reza-rahman if you have any concerns with this approach, please let us know.

**Comment by m-reza-rahman** [comment]
I think this is a very sound approach.

**Comment by kito99** [comment]
@yashTEF, when do you think you'll be able to start working on this? (Also, I invited you tot he slack channel.)

**Comment by yashTEF** [comment]
I was able to make a start and got familiar with the existing header/footer files.

The first part i.e checking whether the new header and footer content causes any problems, seems fairly simple
and should not take much time.

**Comment by yashTEF** [comment]
@autumnfound , The endpoints for header/footer work fine for fetching the html, one minor problem was that the images for logo etc. are being sourced using a relative path instead of absolute:

`images/jakarta/jakarta-ee-logo-color.svg'
`https://jakarta.ee/images/jakarta/jakarta-ee-logo-color.svg'

**Comment by autumnfound** [comment]
> The existing header/footer are also using the absolute path for the images.

This looks like this is a problem specific to this site. I'm submitting a patch for review that should add the host to the URL as there is no need for relative URLs here from what I can tell.

**Comment by autumnfound** [comment]
The change has been merged, so the logo should now have the proper absolute path!

**Comment by kito99** [comment]
@yashTEF do you need to do any more testing, or are is it ready for a final review?

**Comment by yashTEF** [comment]
> @yashTEF do you need to do any more testing, or are is it ready for a final review?

I am working on submitting a PR for fixing the links issue in the header/footer templates to https://github.com/jakartaee/jakarta.ee.

Also the introduction of the new header has displaced the background a bit so will add a commit for that as well, once this is done, it'll be ready for a review.

**Comment by kito99** [comment]
Great, thanks @yashTEF!

**Comment by kito99** [comment]
@yashTEF how is this going?

**Comment by kito99** [comment]
@yashTEF how is it looking for you in the next couple of weeks?

**Comment by kito99** [comment]
@yashTEF just following up..

**Comment by yashTEF** [comment]
Apologies for the delay, haven't logged in for a while.

Seems unlikely that I'll be able to complete this at least until mid next month, any chance that this can be pushed back until then?

**Comment by yashTEF** [comment]
Yes, have started working on this. Will update here on any progress.

**Comment by kito99** [comment]
Hey @yashTEF the month is almost over. Do you think you'll be able to complete this soon?

**Comment by yashTEF** [comment]
Yeah, I have opened a PR: https://github.com/jakartaee/jakarta.ee/pull/1925 which needs to be merged.

Once that is done I think the outstanding PR will be ready for review

**Comment by autumnfound** [comment]
> Yeah, I have opened a PR: [jakartaee/jakarta.ee#1925](https://github.com/jakartaee/jakarta.ee/pull/1925) which needs to be merged.
> 
> Once that is done I think the outstanding PR will be ready for review

We've now merged that change, I hadn't seen the email for the updated review. Should be live in the next 5 minutes!

**Comment by yashTEF** [comment]
Hi @autumnfound , I noticed when fetching the footer from the Jakarta.ee endpoint the
html fragment that I am receiving in response does not have the opening tags as well as the html code for some of the elements for e.g. html, body tags don't have an opening tag.

For inspecting I tried a curl request and viewed the response, Is there any particular reason for this?

**Comment by m-reza-rahman** [comment]
For Faces in particular, valid XHTML would be far preferable.

**Comment by autumnfound** [comment]
> For inspecting I tried a curl request and viewed the response, Is there any particular reason for this?

When it was generated, this was stated as the header and footer of the page that come together to create the base Jakarta page. When combined, the final output would be valid HTML. The intention is that it could be easier to do the string replaces for the placeholders in the header, and attempts to make it less complicated for integration.

> For Faces in particular, valid XHTML would be far preferable.

If it would make it easier, we can look at creating a single HTML document that would be the full page, where the body content would need to be injected into the DOM manually. Should we start tracking a request for that change?

**Comment by kito99** [comment]
Another, less "proper" approach would be to just strip out the </body> and </html> sections from the content returned  for the footer.

**Comment by autumnfound** [comment]
> For inspecting I tried a curl request and viewed the response, Is there any particular reason for this?

When it was generated, this was stated as the header and footer of the page that come together to create the base Jakarta page. When combined, the final output would be valid HTML. The intention is that it could be easier to do the string replaces for the placeholders in the header, and attempts to make it less complicated for integration.

> For Faces in particular, valid XHTML would be far preferable.

If it would make it easier, we can look at creating a single HTML document that would be the full page, where the body content would need to be injected into the DOM manually. We could note in the README what element to inject into, something like a `section` tag with an id of `content` or similar. Should we start tracking a request for that change? 

> Another, less "proper" approach would be to just strip out the and sections from the content returned for the footer.

If you are saying we can include the footer and header sections as valid HTML, this gets really sticky as the content doesn't quite split cleanly, and would likely just introduce more files that would make this even worse to maintain.

**Comment by kito99** [comment]
@autumnfound no, I'm saying you leave it as-is, and we strip out the offending end tags on our end.

**Comment by yashTEF** [comment]
> If it would make it easier, we can look at creating a single HTML document that would be the full page, where the body content would need to be injected into the DOM manually. Should we start tracking a request for that change?

If it makes easier I am able to convert HTML fetched from the url and parse them into well formed XHTML during the build itself using maven plugins.

This works well for the header fragment, but for the footer it removes head and body tags as they don't have matching opening tags, if that can be resolved I should be able to convert the footer HTML to valid XHTML like in the case of header.

**Comment by autumnfound** [comment]
> If it makes easier I am able to convert HTML fetched from the url and parse them into well formed XHTML during the build itself using maven plugins.

If we wanted to use the same "stitching" solution where we give parts of the page to stitch into a larger template, we'd likely have to introduce another 2-3 page fragments at least to get this solution closer to fully functional. There would still likely be some sections where you'd have to hardcode some HTML which would lead to more brittle code, as it would rely on nothing in our styling/page structure changing. Relying on a static structure isn't necessarily the safest thing as it may lead to sudden breaks since you aren't using a static version of the site. 

> This works well for the header fragment, but for the footer it removes head and body tags as they don't have matching opening tags, if that can be resolved I should be able to convert the footer HTML to valid XHTML like in the case of header.

 I've created a [PR](https://gitlab.eclipse.org/eclipsefdn/it/webdev/hugo-solstice-theme/-/merge_requests/371) to add the "full page" template to our base theme. This will need to be reviewed, released, and updated on the Jakarta site, but I believe this will likely be a closer solution to what you need. The change here is that you'll need to update the document model to inject the content into the page. The current query to target will be `div#content-placeholder`, and you can see the base theme version introduced by the PR on this page: https://preview-371--webdev-eclipse-org-docs-hugo.eclipsecontent.org/docs/hugo/templates/page/

**Comment by autumnfound** [comment]
If the above is what you need, let me know and I can start the process of getting this page on the Jakarta EE site. If it doesn't we'll need to figure out something else, as I'm not familiar with Faces as I work with a framework that uses a different templating engine which has it's own quirks

**Comment by autumnfound** [comment]
@yashTEF bumping thread, does the above work, or should we look into something else?

**Comment by yashTEF** [comment]
> @yashTEF bumping thread, does the above work, or should we look into something else?

I am trying to understand the change, how can I use the above to fix the problem?

From what I can understand there would be a single page template instead of header/footer fragments and the page content will be added by manipulating DOM.

@kito99 , does this help or we'll have to change the approach a bit from our side?

**Comment by autumnfound** [comment]
> I am trying to understand the change, how can I use the above to fix the problem?
> 
> From what I can understand there would be a single page template instead of header/footer fragments and the page content will be added by manipulating DOM.

Yep, that's correct! This would eliminate the problem where you have incomplete xhtml fragments which led to some truly strange behaviour where nodes would generate opening tags and lead to bad injection. 

The only other solution I can think of would have something along the lines of the following, which isn't ideal as it would lead to issues if we ever changed the HTML around the main content, and would still require DOM manipulation to fix the title in the metadata of the page in the header section (excuse any pseudo code, I don't use xhtml haha):

```html
<html>
  <xhtml:import uri=".../templates/header" />
  <body>
    <xhtml:import uri=".../templates/content-preamble" />
    <main>
      <div class="container padding-bottom-30">
         <div class="row">
           <div class="col-md-18">
             <div class="default-breadcrumbs hidden-print" id="breadcrumb">
               <div>
                 <div class="row">
                   <div class="col-sm-24">
                     <ol aria-label="Breadcrumb" class="breadcrumb">
                       <li><a href="...">Home</a></li>
                       ... (contextual breadcrumbs here)
                       <li class="active" aria-current="page"><a href="...">{{TITLE}}</a></li>
                     </ol>
                   </div>
                 </div>
               </div>
             </div>
             <div id="content-placeholder">
             
             ...
             
             </div>
           </div>
         </div>
       </div>
     </main>
    <xhtml:import uri=".../templates/footer" />
  </body>
</html>
```

**Comment by autumnfound** [comment]
@yashTEF @kito99 if the mixed implementation I wrote up in pseudo code works better than the other 2 proposed, let me know so I can implement the needed fragments! I'm just waiting on feedback to implement and merge at this point.

**Comment by m-reza-rahman** [comment]
Just to provide a head's up, I will be working to resolve this long-standing issue in the next few weeks.

**Comment by autumnfound** [comment]
> Just to provide a head's up, I will be working to resolve this long-standing issue in the next few weeks.

Sounds good! Let me know what solution works best for you to implement this, as it's still in the state that was previously discussed. Once I know which approach will work better in xhtml, I can push an update to this and we can get it moving again!

**Comment by m-reza-rahman** [comment]
I plan on taking a fresh look to be honest. Ideally I want to see things work with just old fashioned HTML iframe without much JSF involved really. If you have time, we could orient the effort towards that. So if it could be as simple as this it would be great in my view:

```html
<iframe src="https://jakarta.ee/header.html" title="Header"></iframe>
```
```html
<iframe src="https://jakarta.ee/footer.html" title="Footer"></iframe>
```

**Comment by kito99** [comment]
Thanks for taking this on, @m-reza-rahman. Sorry I dropped the ball on this!

**Comment by m-reza-rahman** [comment]
No worries. The main thing is getting this sorted. I myself had to take care of other priorities before I could look at this. As such, we still need to keep an eye on updating the Starter to add support for the EE 11 Web Profile in a timely fashion.

**Comment by autumnfound** [comment]
I've looked into this today, as I wanted to be sure that iframes would be a workable solution. In this case, unfortunately, iframes aren't suitable to the content. We use responsive design in our sites, as well as popover/overlapping elements, with an example being the mobile menu, where it gets cut off if sized properly for normal content, or leaves giant empty gaps if we make space for the menu. I did a quick test locally to check, and here are some examples:

With the mobile menu open
![Image](https://github.com/user-attachments/assets/da870894-6d5d-46b2-928b-271a79ac4b55)

With the "Compatible products" menu open
![Image](https://github.com/user-attachments/assets/b90728f8-1d62-46ae-b873-32100e480619)

If you'd like, I can help by making an update to the xhtml files to get them up to date with current content. In the long term, I think we will need to look at doing some sort of build time script to pull in the sections we need and to create xhtml files, wrapping the content in the `ui:composition` tag. There doesn't appear to be any custom xhtml tags otherwise, so it might fit the need in this case.

**Comment by m-reza-rahman** [comment]
A simple PR just updating the content is always welcome. I don't even think you'll need to muck with JSF for that. Here are the instructions for testing locally: https://github.com/eclipse-ee4j/starter?tab=readme-ov-file#running-the-ui.

Even if can't be iframe, does it need to be done if JSF or could it be a bit more dynamic than that. How about just doing a JavaScipt fetch? So something like this:

```javascript
function fetchHeader() {
  fetch('https://jakarta.ee/header.html')
  .then((response) => {
    return response.text();
  })
  .then((html) => {
    document.header.innerHTML = html     
  });
}
```

**Comment by autumnfound** [comment]
> WDYT?

I think that could potentially work, but there would likely be some flashing, and we'd need to check that none of the JS used to power those sections break from that. We'd want to test that to see if the drawbacks are acceptable

**Comment by m-reza-rahman** [comment]
Do you want to give it a shot with a PR? It would certainly move things along and probably be the least effort in trying to keep things in sync long term. If not, we can use the JSF/compile time approach as a fallback?

**Comment by m-reza-rahman** [comment]
Great! I've merged it - we'll make sure to test and get it out there ASAP.

**Comment by autumnfound** [comment]
> Hi @autumnfound, just something I am thinking about so I figured I'll just ask - if we are the only ones needing this sort of thing, is this really worth making more effort on? How hard would it be just to make a PR with any changes when required?

This project is the only consumer of this feature currently, yeah. In terms of updates, as this is a project-related site, the EF WebDev team is specifically hands off in the case of regular content updates, though we have in the past helped with some set up activities like automation and integrations. So in this case, we can help out in trying to figure out a stable solution for automating the setup of this part of the code, but would leave regular maintenance of the site to the project team.

Just to give an idea of how often this changes, in the past year there have been 5 updates to the navigation menu content, but no large changes to the look and feel in the past few years. The change to bring the content up to date + testing took about 30m in total for me to figure out since I'm used to working with HTML and the browser inspect tool.

**Comment by m-reza-rahman** [comment]
> Just to give an idea of how often this changes, in the past year there have been 5 updates to the navigation menu content, but no large changes to the look and feel in the past few years. The change to bring the content up to date + testing took about 30m in total for me to figure out since I'm used to working with HTML and the browser inspect tool.

Now I am really wondering if we are over-engineering this. Is there a way for me/the project team to find out when these updates happen? We could just pull any requisite changes manually when that happens. This seems the least cumbersome approach really. Anything else is probably just more needless work for everyone involved.

**Comment by autumnfound** [comment]
> Now I am really wondering if we are over-engineering this. Is there a way for me/the project team to find out when these updates happen? We could just pull any requisite changes manually when that happens. This seems the least cumbersome approach really. Anything else is probably just more needless work for everyone involved.

With GitHub, you can watch the repo for Pull requests, but that's about as good as it gets for directly supported solutions. Just as a note, that repo is somewhat busy with 1-3 changes a week on average. There are external tools that watch URLs for content changes, but I've never used them so I can't really give much advice there. 

In terms of what files you care about, watching the [English menu file](https://raw.githubusercontent.com/jakartaee/jakarta.ee/refs/heads/src/config/_default/menu.toml) and the raw content of the 3 files under the [footer_links section](https://github.com/jakartaee/jakarta.ee/tree/src/layouts/partials/footer_links) should give you accurate notices when the content updates!

**Comment by m-reza-rahman** [comment]
This really is the simple and practical approach. I am closing this as resolved.

**Comment by m-reza-rahman** [comment]
Hi @autumnfound, as it turns out, I now have to do this sync: https://github.com/jakartaee/jakarta.ee/pull/2120#event-16978804847. I can of course figure it out, but in the spirit of not reinventing the wheel, can you please share what steps you did to perform the sync? Anything to note or be careful of?
